### PR TITLE
Add TableOfContents component

### LINKED ISSUE #1107: Create table of content component

### LINKED ISSUE #1108: Make table of content component selectable

### LINKED ISSUE #1109: Entries QA: Transform framework data to be supported by ToC

### PR DESCRIPTION
**Pull request by samshara**
fixes https://github.com/the-deep/client/issues/1107
fixes https://github.com/the-deep/client/issues/1108
fixes https://github.com/the-deep/client/issues/1109
fix styling

### CODE REVIEW

File: src/views/Entries/QualityControl/index.tsx [hunk 1]
```diff
@@ -1,29 +1,77 @@
...
     const {
         className,
         // projectId,
     } = props;
 
+    const [selected, setSelection] = useState<number | string | undefined>(undefined);
+
+    const handleSelection = (id: number | string) => {
+        setSelection(id);
+    };
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/index.tsx]: Why do we have type for `selected` as `string | undefined`
shouldn't be be either `string` or a `number`

File: src/components/TableOfContents/index.tsx [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+            case 2: {
+                return styles.levelTwo;
+            }
+            case 3: {
+                return styles.levelThree;
+            }
+            default: {
+                return styles.leaf;
+            }
+        }
```
Comment (tnagorra) [inline src/components/TableOfContents/index.tsx]: I guess we should use a map here
```javascript
return styleMapping[level] ?? styles.leaf;
```

File: src/components/TableOfContents/index.tsx [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    };
+
+    const isSelected = id === value;
+
+    const levelClassName = getClassName();
```
Comment (tnagorra) [inline src/components/TableOfContents/index.tsx]: Why not just use useMemo insetad of creating a function here?

File: src/components/TableOfContents/index.tsx [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+
+    const handleClick = () => {
+        onChange(id);
+    };
+
+    const [collapsed, setCollapsed] = useState<boolean>(false);
+
+    const handleCollapseToggle = () => {
+        setCollapsed(v => !v);
+    };
```
Comment (tnagorra) [inline src/components/TableOfContents/index.tsx]: use useCallback

File: src/components/TableOfContents/index.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        childrenSelector,
+    } = props;
+
+    const id = keySelector(options);
+    const title = labelSelector(options);
+    const children = childrenSelector(options);
+
+    const handleClick = () => {
+        onChange(id);
+    };
```
Comment (tnagorra) [inline src/components/TableOfContents/index.tsx]: use useCallback

File: src/components/TableOfContents/index.tsx [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+import styles from './styles.css';
+
+interface Props<T>{
+    keySelector: (datum: T) => string | number;
+    labelSelector: (datum: T) => string | number;
+    childrenSelector: (datum: T) => T[] | undefined;
+    onChange: (key: string | number) => void;
+    options: T[];
+    value: string | number | undefined;
+}
```
Comment (tnagorra) [inline src/components/TableOfContents/index.tsx]: I think we should have a generic type for key.

```suggestion
interface Props<T, K extends string | number>{
    keySelector: (datum: T) => string | number;
    labelSelector: (datum: T) => K;
    childrenSelector: (datum: T) => T[] | undefined;
    onChange: (key: K) => void;
    options: T[];
    value: K | undefined;
}
```

File: src/utils/framework.ts [hunk 7]
```diff
@@ -256,3 +257,122 @@ export const getMatrix2dStructures = (framework: MiniFrameworkElement | undefine
...
+    }
+
+    const {
+        widgets,
+    } = framework;
+
+    const matrix1dList = widgets.filter(d => d.widgetId === 'matrix1dWidget');
+
+    if (matrix1dList.length === 0) {
+        return emptyObject;
```
Comment (tnagorra) [inline src/utils/framework.ts]: Shouldn't be be emptyArray?

Comment (tnagorra) [reply src/utils/framework.ts]: We can completely skip this condition and get an emptyArray as return value

File: src/utils/framework.ts [hunk 8]
```diff
@@ -256,3 +257,122 @@ export const getMatrix2dStructures = (framework: MiniFrameworkElement | undefine
...
+            children: transformedRows,
+        });
+    });
+
+    return toc;
+};
+
+export const getMatrix2dToc = (framework: MiniFrameworkElement | undefined) => {
+    if (!framework) {
+        return emptyObject;
```
Comment (tnagorra) [inline src/utils/framework.ts]: This should be empty array

File: src/utils/framework.ts [hunk 9]
```diff
@@ -256,3 +257,122 @@ export const getMatrix2dStructures = (framework: MiniFrameworkElement | undefine
...
+
+    const {
+        widgets,
+    } = framework;
+
+    const matrix2dList = widgets.filter(d => d.widgetId === 'matrix2dWidget');
+
+    if (matrix2dList.length === 0) {
+        return emptyObject;
+    }
```
Comment (tnagorra) [inline src/utils/framework.ts]: we can skip this check

File: src/views/Entries/QualityControl/index.tsx [hunk 10]
```diff
@@ -64,8 +48,22 @@ function QualityControl(props: QualityControlProps) {
                 leftContainerClassName={styles.left}
                 leftChild={(
                     <div className={styles.frameworkSelection}>
+                        <div>
+                            Matrix 1d
+                        </div>
+                        <TableOfContents
+                            options={matrix1dToc as MatrixTocElement[]}
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/index.tsx]: We shouldn't cast the type of matrix1dToc, instead the return type for getMatrix1dToc should be MatrixTocElement[]

File: src/views/Entries/QualityControl/index.tsx [hunk 11]
```diff
@@ -64,8 +48,22 @@ function QualityControl(props: QualityControlProps) {
...
+                            childrenSelector={childrenSelector}
+                            onChange={handleSelection}
+                            value={selected}
+                        />
+                        <div>
+                            Matrix 2d
+                        </div>
                         <TableOfContents
-                            options={toc}
+                            options={matrix2dToc as MatrixTocElement[]}
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/index.tsx]: We shouldn't cast the type of matrix2dToc, instead the return type for getMatrix2dToc should be MatrixTocElement[]

File: src/views/Entries/QualityControl/index.tsx [hunk 12]
```diff
@@ -1,29 +1,69 @@
...
     return (
         <div className={_cs(className, styles.qualityControl)}>
             <ResizableH
                 className={styles.resizableContainer}
                 leftContainerClassName={styles.left}
                 leftChild={(
                     <div className={styles.frameworkSelection}>
-                        Framework selection
+                        <div>
+                            Matrix 1d
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/index.tsx]: Translate here

File: src/views/Entries/QualityControl/index.tsx [hunk 13]
```diff
@@ -1,29 +1,69 @@
...
+                        <TableOfContents
+                            options={matrix1dToc}
+                            keySelector={keySelector}
+                            labelSelector={labelSelector}
+                            childrenSelector={childrenSelector}
+                            onChange={handleSelection}
+                            value={selected}
+                        />
+                        <div>
+                            Matrix 2d
```
Comment (tnagorra) [inline src/views/Entries/QualityControl/index.tsx]: Translate here

File: src/utils/framework.ts [hunk 14]
```diff
@@ -256,3 +255,144 @@ export const getMatrix2dStructures = (framework: MiniFrameworkElement | undefine
...
+            children: transformedRows,
+        });
+    });
+
+    return toc;
+};
+
+export const getMatrix2dToc = (framework: MiniFrameworkElement | undefined): MatrixTocElement[] => {
+    if (!framework) {
+        return [];
```
Comment (tnagorra) [inline src/utils/framework.ts]:
```suggestion
        return emptyArray;
```

File: src/components/TableOfContents/index.tsx [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+
+    const id = keySelector(options);
+    const title = labelSelector(options);
+    const children = childrenSelector(options);
+
+    const handleClick = useCallback(() => onChange(id), [id, onChange]);
+
+    const [collapsed, setCollapsed] = useState<boolean>(false);
+
+    const handleCollapseToggle = useCallback(() => setCollapsed(v => !v), []);
```
Comment (tnagorra) [inline src/components/TableOfContents/index.tsx]:
```suggestion
    const handleCollapseToggle = useCallback(
        () => setCollapsed(v => !v),
        [],
    );
```

File: src/components/TableOfContents/index.tsx [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+        keySelector,
+        labelSelector,
+        childrenSelector,
+    } = props;
+
+    const id = keySelector(options);
+    const title = labelSelector(options);
+    const children = childrenSelector(options);
+
+    const handleClick = useCallback(() => onChange(id), [id, onChange]);
```
Comment (tnagorra) [inline src/components/TableOfContents/index.tsx]:
```suggestion
    const handleClick = useCallback(
    () => onChange(id),
     [id, onChange],
 );
```

**Review by frozenhelium — APPROVED** [review]

**Pull request merged**